src/masternodes/mn_checks.cpp

```diff
@@ -398,6 +447,54 @@ class CCustomTxVisitor : public boost::static_visitor<Res>
         return Res::Ok();
     }
 
+    Res CheckICXTx() const {
+        if (tx.vout.size() != 2) {
+            return Res::Err("malformed tx vouts ((wrong number of vouts)");
+        }
+        return Res::Ok();
+    }
+
+    Res ICXTransfer(DCT_ID id, CAmount amount, CScript const & from, CScript const & to) const {
+        assert(!from.empty() || !to.empty());
+
+        CTokenAmount tokenAmount{id, amount};
+        // if "from" not supplied it will only add balance on "to" address
+        if (!from.empty()) {
+            auto res = mnview.SubBalance(from, tokenAmount);
+            if (!res)
+                return res;
+        }
+
+        // if "to" not supplied it will only sub balance from "form" address
+        if (!to.empty()) {
+            auto res = mnview.AddBalance(to,tokenAmount);
+            if (!res)
+                return res;
+        }
+
+        return Res::Ok();
+    }
+
+    DCT_ID FindTokenByPartialSymbolName(const std::string& symbol) const {
+        DCT_ID res{0};
+        mnview.ForEachToken([&](DCT_ID id, CTokenImplementation token) {
+            if (token.symbol.find(symbol) == 0) {
+                res = id;
+                return false;
+            }
+            return true;
+        }, DCT_ID{1});
+        assert(res.v != 0);
+        return res;
+    }
+
+    CPoolPair GetDFIBTCPoolPair() const {
+        auto BTC = FindTokenByPartialSymbolName(CICXOrder::TOKEN_BTC);
+        auto pair = mnview.GetPoolPair(DCT_ID{0}, BTC);
+        assert(pair);
```

Comment:
Can return some fail here? assert is not compiled in release mode

Reply:
BTC does not exist on regtest, but we should have tests create BTC in the beginning. But if someone trys to do something without creating BTC token it will fail.

Reply:
Sure, if BTC does not exist it's an unusual env. Assert enforces correct env.